common/persistence/nosql/nosqlplugin/cassandra/queue.go

```diff
@@ -0,0 +1,255 @@
+	previous := make(map[string]interface{})
+	applied, err := query.MapScanCAS(previous)
+	if err != nil {
+		return err
+	}
+
+	if !applied {
+		return errConditionFailed
+	}
+	return nil
+}
+
+// Get the ID of last message inserted into the queue
+func (db *cdb) GetLastEnqueuedMessageID(ctx context.Context, queueType persistence.QueueType) (int64, error) {
+	query := db.session.Query(templateGetLastMessageIDQuery, queueType).WithContext(ctx)
+	result := make(map[string]interface{})
+	err := query.MapScan(result)
+	if err != nil {
+		return 0, err
+	}
+
+	return result["message_id"].(int64), nil
+}
+
+// Read queue messages starting from the exclusiveBeginMessageID
+func (db *cdb) GetMessagesFromQueue(
+	ctx context.Context,
+	queueType persistence.QueueType,
+	exclusiveBeginMessageID int64,
+	maxRows int,
+) ([]*nosqlplugin.QueueMessageRow, error) {
+	// Reading replication tasks need to be quorum level consistent, otherwise we could loose task
+	query := db.session.Query(templateGetMessagesQuery,
+		queueType,
+		exclusiveBeginMessageID,
+		maxRows,
+	).WithContext(ctx)
+
+	iter := query.Iter()
+	if iter == nil {
+		return nil, fmt.Errorf("GetMessagesFromQueue operation failed. Not able to create query iterator")
+	}
+
+	var result []*nosqlplugin.QueueMessageRow
+	message := make(map[string]interface{})
+	for iter.MapScan(message) {
+		payload := getMessagePayload(message)
+		id := getMessageID(message)
+		result = append(result, &nosqlplugin.QueueMessageRow{ID: id, Payload: payload})
+		message = make(map[string]interface{})
+	}
+
+	if err := iter.Close(); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// Read queue message starting from exclusiveBeginMessageID int64, inclusiveEndMessageID int64
+func (db *cdb) GetMessagesBetween(
+	ctx context.Context,
+	request nosqlplugin.GetMessagesBetweenRequest,
+) (*nosqlplugin.GetMessagesBetweenResponse, error) {
+	// Reading replication tasks need to be quorum level consistent, otherwise we could loose task
+	// Use negative queue type as the dlq type
+	query := db.session.Query(templateGetMessagesFromDLQQuery,
+		request.QueueType,
+		request.ExclusiveBeginMessageID,
+		request.InclusiveEndMessageID,
+	).PageSize(request.PageSize).PageState(request.NextPageToken).WithContext(ctx)
+
+	iter := query.Iter()
+	if iter == nil {
+		return nil, fmt.Errorf("GetMessagesBetween operation failed. Not able to create query iterator")
+	}
+
+	var rows []nosqlplugin.QueueMessageRow
+	message := make(map[string]interface{})
+	for iter.MapScan(message) {
+		payload := getMessagePayload(message)
+		id := getMessageID(message)
+		rows = append(rows, nosqlplugin.QueueMessageRow{ID: id, Payload: payload})
+		message = make(map[string]interface{})
+	}
+
+	nextPageToken := iter.PageState()
+	// TODO? can we remove the below code 
+	//newPageToken := make([]byte, len(nextPageToken))
+	//copy(newPageToken, nextPageToken)
```

Comment:
@meiliang86 the existing code copy the token to a new slice, I don't think that's necessary, but let me know if I missed something.